cpp/src/arrow/array/array_base.h

```diff
@@ -165,6 +166,28 @@ class ARROW_EXPORT Array {
   /// An error is returned if the types are not layout-compatible.
   Result<std::shared_ptr<Array>> View(const std::shared_ptr<DataType>& type) const;
 
+  /// \brief Construct a copy of the array with all buffers on destination
+  /// Memory Manager
+  ///
+  /// This method recursively copies the array's buffers and those of its children
+  /// onto the destination MemoryManager device and returns the new Array.
+  Result<std::shared_ptr<Array>> CopyTo(const std::shared_ptr<MemoryManager>& to) const {
+    ARROW_ASSIGN_OR_RAISE(auto copied_data, data()->CopyTo(to));
```

Comment:
I don't think there's a reason to inline this function definition in the `.h` file. By moving it into the corresponding `.cc`, we can cut down on an additional header inclusion.